### ISSUE TITLE
tmp files not cleaned with sudo_user

### ISSUE THREAD

**Issue by albin-hubert** [ansible/ansible#2693]
When using sudo, the tmp files are created in /tmp/ using the ssh user.

However for non root sudo, the "rm -fr /tmp/..." is executed through the sudo using the "-u sudo_user". It then fails due to permission and the tmp files are not removed.

```
    ESTABLISH CONNECTION FOR USER: ansible on PORT 22 TO xxxx
    EXEC /bin/sh -c 'mkdir -p /tmp/ansible-1366115679.65-230877770548883 && chmod a+rx /tmp/ansible-1366115679.65-230877770548883 && echo /tmp/ansible-1366115679.65-230877770548883'
    REMOTE_MODULE ping
    PUT /tmp/tmpT7QiSY TO /tmp/ansible-1366115679.65-230877770548883/ping
    EXEC /bin/sh -c 'chmod a+r /tmp/ansible-1366115679.65-230877770548883/ping'
    EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=....] password: " -u alfresco /bin/sh -c '"'"' /usr/bin/python /tmp/ansible-1366115679.65-230877770548883/ping; rm -rf /tmp/ansible-1366115679.65-230877770548883/ >/dev/null 2>&1'"'"''
```

The last command executed without the error redirect to /dev/null:

```
    /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=....] password: " -u alfresco /bin/sh -c '"'"' /usr/bin/python /tmp/ansible-1366115679.65-230877770548883/ping; rm -rf /tmp/ansible-1366115679.65-230877770548883/'"'"''
    {"changed": false, "ping": "pong"}
    rm: cannot remove `/tmp/ansible-1366115679.65-230877770548883//ping': Permission denied
```

**Comment by mpdehaan** [comment]
Please supply more output about the command you ran and the playbook you used.  I feel I'm looking through a porthole here and not seeing everything.

This hasn't been reported or hanged recently, so something weird seems to be going on.

Can you show me the output of ls -l and such on the remote system of this particular file?

**Comment by albin-hubert** [comment]
It is really not so complicated.
Here is the full example:

Playbook

```
   ---
   - name: example
     hosts: whatever
     user: ansible
     gather_facts: no
     sudo: True
     sudo_user: alfresco
     tasks:
        - name: ping
          action: ping
```

Launching the playbook gives: 

```
   ]# ansible-playbook -vvv ping.yml

   PLAY [example] *********************

   TASK: [ping] *********************
   <whatever> ESTABLISH CONNECTION FOR USER: ansible on PORT 22 TO whatever
   <whatever> EXEC /bin/sh -c 'mkdir -p /tmp/ansible-1366187718.03-229895154216961 && chmod a+rx /tmp/ansible-1366187718.03-229895154216961 && echo /tmp/ansible-1366187718.03-229895154216961'
   <whatever> REMOTE_MODULE ping
   <whatever> PUT /tmp/tmpwIIVAD TO /tmp/ansible-1366187718.03-229895154216961/ping
   <whatever> EXEC /bin/sh -c 'chmod a+r /tmp/ansible-1366187718.03-229895154216961/ping'
   <whatever> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=lanffuagsiihngsndhdhvothgtchevqd] password: " -u alfresco /bin/sh -c '"'"' /usr/bin/python /tmp/ansible-1366187718.03-229895154216961/ping; rm -rf /tmp/ansible-1366187718.03-229895154216961/ >/dev/null 2>&1'"'"''
   ok: [whatever] => {"changed": false, "ping": "pong"}

   PLAY RECAP *********************
   whatever                     : ok=1    changed=0    unreachable=0    failed=0
```

Tmp files have not been cleaned up:

```
   ]# ll /tmp/ansible-1366187718.03-229895154216961/
   -rw-rw-r-- 1 ansible ansible 31366 Apr 17 10:35 ping
```

The important part are:
- Here you are creating the tmp directory using the ssh user (in my case ansible)
  
     <whatever> EXEC /bin/sh -c 'mkdir -p /tmp/ansible-1366187718.03-229895154216961 && chmod a+rx /tmp/ansible-1366187718.03-229895154216961 && echo /tmp/ansible-1366187718.03-229895154216961'
- Then you are trying to remove it by using the sudo user (in my case alfresco) instead of the real owner of the file (ansible)
  
     <whatever> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=lanffuagsiihngsndhdhvothgtchevqd] password: " -u alfresco /bin/sh -c '"'"' /usr/bin/python /tmp/ansible-1366187718.03-229895154216961/ping; rm -rf /tmp/ansible-1366187718.03-229895154216961/ >/dev/null 2>&1'"'"''

The problem comes from the facts that the rm -fr command is executed through the sudo at the same time as the python module. You should run the rm -fr command after without sudoing it.

**Comment by mpdehaan** [comment]
Ah, so when the sudo_user is _not_ root, we need to do things in two steps,
got it.

I'll check.

**Comment by mpdehaan** [comment]
Fix pushed in 9871707 ... let me know if this works for you, thanks!
